reframe/frontend/cli.py

```diff
@@ -111,6 +111,124 @@ def generate_report_filename(filepatt):
     return filepatt.format(sessionid=new_id)
 
 
+def filter_checks(checks_found, options, rt, loader, ci_tag=None):
+
+    # Filter checks by name
+    checks_matched = checks_found
+    if options.exclude_names:
+        for name in options.exclude_names:
+            checks_matched = filter(filters.have_not_name(name),
+                                    checks_matched)
+
+    if options.names:
+        checks_matched = filter(filters.have_name('|'.join(options.names)),
+                                checks_matched)
+
+    # Filter checks by tags
+    for tag in options.tags:
+        checks_matched = filter(filters.have_tag(tag), checks_matched)
+
+    # Filter checks by ci tags
+    if ci_tag:
+        checks_matched = filter(filters.have_tag(ci_tag), checks_matched)
+
+    # Filter checks by prgenv
+    if not options.skip_prgenv_check:
+        for prgenv in options.prgenv:
+            checks_matched = filter(filters.have_prgenv(prgenv),
+                                    checks_matched)
+
+    # Filter checks by system
+    if not options.skip_system_check:
+        checks_matched = filter(
+            filters.have_partition(rt.system.partitions), checks_matched)
+
+    if options.gpu_only:
+        checks_matched = filter(filters.have_gpu_only(), checks_matched)
+    elif options.cpu_only:
+        checks_matched = filter(filters.have_cpu_only(), checks_matched)
+
+    # Determine the allowed programming environments
+    allowed_environs = {e.name
+                        for env_patt in options.prgenv
+                        for p in rt.system.partitions
+                        for e in p.environs if re.match(env_patt, e.name)}
+
+    # Generate the test cases, validate dependency and sort them
+    checks_matched = list(checks_matched)
+
+    return checks_matched
+
+# TODO place this function in a proper module
+# TODO define mechanism on how to propagate command line options, currently
+#      we are passing site_config, but that's clearly a bad design
+#      read below the options we need to propagate or make available to each
+#      ci invokation
+#      An alternative design would be to have environment variables exported
+#      during the trigger phase of reframe that would influence the tests
+#      individually
+def generate_ci_pipeline_file(ci_pipeline_file, ci_tags, checks_found, options, rt, loader, site_config):
```

Comment:
All the CI integration functionality must go into a separate module.

Reply:
And the interface seems a bit hacky, too.

Reply:
The reason is because this function mixes everything: selection of tests and the generation of the pipeline. These must be separated and, especially, the selection of the tests must be viewed in the light of the work done in #1594.